File: kotlin-native/backend.native/tests/objcexport/overrideKotlinMethods.swift
/*
 * Copyright 2010-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
 * that can be found in the LICENSE file.
 */

import Kt

class OverrideKotlinMethodsImpl : OverrideKotlinMethods4, OverrideKotlinMethods6 {
    override func one() -> Int32 {
        return 1
    }
}

private func test1() throws {
    let obj = OverrideKotlinMethodsImpl()

    try OverrideKotlinMethodsKt.test0(obj: obj)
    try OverrideKotlinMethodsKt.test1(obj: obj)
    try OverrideKotlinMethodsKt.test2(obj: obj)
    try OverrideKotlinMethodsKt.test3(obj: obj)
    try OverrideKotlinMethodsKt.test4(obj: obj)
    try OverrideKotlinMethodsKt.test5(obj: obj)
    try OverrideKotlinMethodsKt.test6(obj: obj)
}

class OverrideKotlinMethodsTests : SimpleTestProvider {
    override init() {
        super.init()

        test("Test1", test1)
    }
}

File: native/native.tests/testData/framework/objcexport/overrideKotlinMethods.kt
package overrideKotlinMethods

import kotlin.test.*

internal interface OverrideKotlinMethods0<T> {
    fun one(): T
}

internal interface OverrideKotlinMethods1<T> : OverrideKotlinMethods0<T>

interface OverrideKotlinMethods2 {
    fun one(): Int
}

open class OverrideKotlinMethods3 {
    internal open fun one(): Number = 3
}

open class OverrideKotlinMethods4 : OverrideKotlinMethods3(), OverrideKotlinMethods1<Int>, OverrideKotlinMethods2 {
    override fun one(): Int = 2
}

interface OverrideKotlinMethods5 {
    fun one(): Int
}

interface OverrideKotlinMethods6 : OverrideKotlinMethods5

// Using `Any` because Kotlin forbids internal type in public function signature.
@Throws(Throwable::class)
fun test0(obj: Any) {
    val obj0 = obj as OverrideKotlinMethods0<*>
    assertEquals(1, obj0.one())
}

// Using `Any` because Kotlin forbids internal type in public function signature.
@Throws(Throwable::class)
fun test1(obj: Any) {
    val obj1 = obj as OverrideKotlinMethods1<*>
    assertEquals(1, obj1.one())
}

@Throws(Throwable::class)
fun test2(obj: OverrideKotlinMethods2) {
    assertEquals(1, obj.one())
}

@Throws(Throwable::class)
fun test3(obj: OverrideKotlinMethods3) {
    assertEquals(1, obj.one())
}

@Throws(Throwable::class)
fun test4(obj: OverrideKotlinMethods4) {
    assertEquals(1, obj.one())
}

@Throws(Throwable::class)
fun test5(obj: OverrideKotlinMethods5) {
    assertEquals(1, obj.one())
}

@Throws(Throwable::class)
fun test6(obj: OverrideKotlinMethods6) {
    assertEquals(1, obj.one())
}
